sklearn/ensemble/gradient_boosting_cv.py

```diff
@@ -0,0 +1,804 @@
+            'max_features': self.max_features,
+            'max_leaf_nodes': self.max_leaf_nodes,
+            'presort': self.presort,
+            'random_state': self.random_state,
+        }
+
+        # Pre processing to ensure every parameter is a list
+        # which `ParameterGrid` can iterate over
+        for key, value in params.items():
+            params[key] = np.atleast_1d(value)
+
+        return params
+
+
+def _fit_single_param(estimator, X, y, train, validation, params, stop_rounds,
+                      max_iter, scoring, score_precision):
+    """ Fit a single estimator till stopping criteria is reached.
+
+    Parameters
+    ---------
+    estimator : GradientBoostingRegressor or GradientBoostingClassifier
+        The class of the model to instantiate.
+
+    X : array-like, shape = [n_samples, n_features]
+        Training vectors, where n_samples is the number of samples
+        and n_features is the number of features.
+
+    y : array-like, shape = [n_samples]
+        Target values (integers in classification, real numbers in
+        regression)
+        For classification, labels must correspond to classes.
+
+    train : array-like
+        The indices of the training samples.
+
+    validation : array-like
+        The indices of the validation samples.
+
+    params : dict
+        The parameters to instantiate the Gradient Boosting model with
+
+    stop_rounds : int
+        The number of iterations for which the score should remain constant
+        over to stop training.
+
+    max_iter : int
+        The maximum number of estimators to fit during gradient boosting.
+
+    scoring : string, callable or None, default=None
+        A string (see model evaluation documentation) or
+        a scorer callable object / function with signature
+        ``scorer(estimator, X, y)``.
+        If ``None``, the ``score`` method of the estimator is used.
+
+    score_precision : int
+        The number of decimal places the score during each iterations is
+        rounded off to before comparing with the previous ones.
+
+    Returns
+    -------
+    score: float
+        The score when the model converged.
+
+    n_iterations:
+        The number of iterations it took to converge. This is the same as
+        parameter `n_estimators` for the gradient boosting model
+    """
+
+    params['warm_start'] = True
+    gb = estimator(**params)
+    scorer = check_scoring(estimator, scoring=scoring)
+    scores = np.empty((stop_rounds,))
+    scores.fill(-np.inf)
+
+    X_train = X[train]
+    y_train = y[train]
+    X_validation = X[validation]
+    y_validation = y[validation]
+
+    for i in range(1, max_iter):
+        gb.n_estimators = i
+        gb.fit(X_train, y_train)
+
+        scores = np.roll(scores, shift=-1)
+        scores[-1] = scorer(gb, X_validation, y_validation)
+        rounded_scores = np.round(scores, score_precision)
+
+        if np.all(rounded_scores[-1] <= rounded_scores):
```

Comment:
But that assumes that all the scores are strictly decreasing or same right? Would there be no case where one score drops spuriously?

Say

```
  0     1     2   3     4     5     6     7     8     9    10
0.91, 0.8, 0.9, 0.89, 0.88, 0.87, 0.86, 0.85, 0.84, 0.83, 0.82
```

In this case we don't want to stop at 10 because 10 is not better than 1 correct?